pkg/compactor/compactor.go

```diff
@@ -0,0 +1,200 @@
+
+	if len(ro.DeltaSnapList) > 0 {
+		cp.logger.Infof("Applying delta snapshots...")
+		if err := r.ApplyDeltaSnapshots(client, *ro); err != nil {
+			cp.logger.Warnf("Could not apply the delta snapshots: %v", err)
+		}
+	} else {
+		cp.logger.Infof("No delta snapshots present over base snapshot.")
+	}
+
+	// Then compact ETCD
+	ctx := context.TODO()
+	revCheckCtx, cancel := context.WithTimeout(ctx, etcdDialTimeout)
+	getResponse, err := client.Get(revCheckCtx, "foo")
+	cancel()
+	if err != nil {
+		return nil, fmt.Errorf("failed to connect to client: %v", err)
+	}
+	etcdRevision := getResponse.Header.GetRevision()
+
+	client.Compact(ctx, etcdRevision)
+
+	// Then defrag the ETCD
+	if needDefragmentation {
+		var dbSizeBeforeDefrag, dbSizeAfterDefrag int64
+		statusReqCtx, cancel := context.WithTimeout(ctx, etcdDialTimeout)
+		status, err := client.Status(statusReqCtx, ep[0])
+		cancel()
+		if err != nil {
+			cp.logger.Warnf("Failed to get status of etcd member[%s] with error: %v", ep, err)
+		} else {
+			dbSizeBeforeDefrag = status.DbSize
+		}
+
+		start := time.Now()
+		defragCtx, cancel := context.WithTimeout(ctx, time.Duration(60*time.Second))
+		_, err = client.Defragment(defragCtx, ep[0])
+		cancel()
+		if err != nil {
+			cp.logger.Errorf("Total time taken to defragment: %v", time.Now().Sub(start).Seconds())
+			cp.logger.Errorf("Failed to defragment etcd member[%s] with error: %v", ep, err)
+		}
+		cp.logger.Infof("Total time taken to defragment: %v", time.Now().Sub(start).Seconds())
+		cp.logger.Infof("Finished defragmenting etcd member[%s]", ep)
+
+		statusReqCtx, cancel = context.WithTimeout(ctx, etcdDialTimeout)
+		status, err = client.Status(statusReqCtx, ep[0])
+		cancel()
+		if err != nil {
+			cp.logger.Warnf("Failed to get status of etcd member[%s] with error: %v", ep, err)
+		} else {
+			dbSizeAfterDefrag = status.DbSize
+			cp.logger.Infof("Probable DB size change for etcd member [%s]:  %dB -> %dB after defragmentation", ep, dbSizeBeforeDefrag, dbSizeAfterDefrag)
+		}
+	}
+
+	// Then take snapeshot of ETCD
+	snapshotReqCtx, cancel := context.WithTimeout(ctx, etcdDialTimeout)
+	defer cancel()
+
+	rc, err := client.Snapshot(snapshotReqCtx)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to create etcd snapshot out of compacted DB: %v", err)
+	}
+
+	isCompressed, compressionPolicy, err := compressor.IsSnapshotCompressed(ro.BaseSnapshot.CompressionSuffix)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to determine if snapshot is compressed: %v", ro.BaseSnapshot.CompressionSuffix)
+	}
+	if isCompressed {
+		rc, err = compressor.CompressSnapshot(rc, compressionPolicy)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to obtain reader for compressed file: %v", err)
+		}
+	}
+	defer rc.Close()
+
+	cp.logger.Infof("Successfully opened snapshot reader on etcd")
+
+	// Then save the snapshot to the store.
+	s := snapstore.NewSnapshot(snapstore.SnapshotKindFull, 0, etcdRevision, ro.BaseSnapshot.CompressionSuffix)
```

Comment:
Won't this upload the compacted snapshot to the same snapstore where it was restored from? Won't that be a bit dangerous if someone runs this command manually? We had discussed saving it locally in the current working directory, right?

Reply:
Updated the PR with the necessary changes